File: go-solution/UserService/services/user_service.go
package services

import (
	"fmt"

	"github.com/albertmakan/scipaper.io/go-solution/UserService/dto"
	"github.com/albertmakan/scipaper.io/go-solution/UserService/helpers"
	"github.com/albertmakan/scipaper.io/go-solution/UserService/models"
	"github.com/albertmakan/scipaper.io/go-solution/UserService/repository"
)

type UserService struct {
	userRepository *repository.UserRepository
}

func NewUserService(userRepository *repository.UserRepository) *UserService {
	return &UserService{userRepository: userRepository}
}

func (userService *UserService) Register(user* models.User) error {
	if userService.FindByUsername(user.Username) != nil {
		return fmt.Errorf("user with username %v already exists", user.Username)
	}
	user.Password, _ = helpers.HashPassword(user.Password)
	userService.userRepository.Create(user)
	return nil
}

func (userService *UserService) GetAll() *[]models.User {
	return userService.userRepository.GetAll()
}

func (userService *UserService) FindByUsername(username string) *models.User {
	return userService.userRepository.FindByUsername(username)
}

func (userService *UserService) Authenticate(username, password string) (*dto.AuthenticatedUser, error) {
	user := userService.FindByUsername(username)
	if user == nil {
		return nil, fmt.Errorf("invalid username or password")
	}
	if !helpers.CheckPasswordHash(password, user.Password) {
		return nil, fmt.Errorf("invalid username or password")
	}
	claims := helpers.Claims{
		Name: username,
	}
	jwt, _ := helpers.GetJwtToken(claims)
	return &dto.AuthenticatedUser{Jwt:jwt}, nil
}

File: dotnet-solution/ScipaperIO/UserService/Services/Impl/UserService.cs
using UserService.Exceptions;
using UserService.Models;
using UserService.Repository.Contracts;
using UserService.Services.Base;
using BCryptNet = BCrypt.Net.BCrypt;

namespace UserService.Services.Impl
{
    public class UserService : IUserService
    {
        private readonly IUserRepository _userRepository;
        private readonly ITokenProvider _tokenProvider;

        public UserService(IUserRepository repository, ITokenProvider tokenProvider)
        {
            _userRepository = repository;
            _tokenProvider = tokenProvider;
        }

        public string Authenticate(string username, string password)
        {
            var user = _userRepository.FindByUsername(username);
            if (user == null)
                throw new UnauthorizedException("invalid username or password");
            if (!BCryptNet.Verify(password, user.Password))
                throw new UnauthorizedException("invalid username or password");
            return _tokenProvider.GenerateToken(username);
        }

        public User CreateUser(User user)
        {
            if (_userRepository.FindByUsername(user.Username) != null)
                throw new BadLogicException("Username already exists");
            user.Password = BCryptNet.HashPassword(user.Password);
            return _userRepository.InsertOne(user);
        }

        public User GetUser(string id)
        {
            return _userRepository.FindById(id);
        }

        public string GetUserName(string token)
        {
            string username = token;
            var user = _userRepository.FindByUsername(username);
            if (user == null) return "";
            return $"{username} {user.FirstName} {user.LastName}";
        }

        public bool IsAuthenticated(string token)
        {
            return true;
        }
    }
}
